File: packages/analysis/test/snowball-fixture.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { gunzipSync } from "node:zlib";

export function readSnowballFixture(language: string): {
  words: string[];
  stems: string[];
} {
  const fixtureRoot = join(
    import.meta.dirname,
    "..",
    "..",
    "..",
    "spec",
    "fixtures",
    "snowball",
  );
  const readLines = (kind: "input" | "output") =>
    gunzipSync(readFileSync(join(fixtureRoot, `${language}-${kind}.txt.gz`)))
      .toString("utf8")
      .trim()
      .split(/\r?\n/);

  return { words: readLines("input"), stems: readLines("output") };
}

export function collectStemMismatches(
  words: readonly string[],
  stems: readonly string[],
  stem: (word: string) => string,
): string[] {
  const mismatches: string[] = [];
  for (let index = 0; index < words.length; index++) {
    const word = words[index] as string;
    const expected = stems[index] as string;
    const actual = stem(word);
    if (actual !== expected) {
      mismatches.push(`${word}: expected "${expected}", got "${actual}"`);
    }
  }
  return mismatches;
}

File: python/searchable-analysis/tests/snowball_fixture.py
import gzip
from collections.abc import Callable
from pathlib import Path

FIXTURES = Path(__file__).parents[3] / "spec" / "fixtures" / "snowball"


def assert_snowball_vocabulary(language: str, stem: Callable[[str], str]) -> None:
    with gzip.open(FIXTURES / f"{language}-input.txt.gz", "rt", encoding="utf-8") as source:
        words = source.read().splitlines()
    with gzip.open(FIXTURES / f"{language}-output.txt.gz", "rt", encoding="utf-8") as source:
        expected = source.read().splitlines()

    assert len(words) == len(expected)
    assert [stem(word) for word in words] == expected
